charms/istio-gateway/src/charm.py

```diff
@@ -40,17 +24,22 @@ def __init__(self, *args):
     def install(self, event):
         """Install charm."""
 
+        if not self.unit.is_leader():
+            return self.unit, WaitingStatus("Waiting for leadership")
+
+        try:
+            self.interfaces = get_interfaces(self)
+        except NoVersionsListed as err:
+            return self.app, WaitingStatus(str(err))
+
         if self.model.config['kind'] not in ('ingress', 'egress'):
-            self.model.unit.status = BlockedStatus('Config item `kind` must be set')
-            return
+            return self.app, BlockedStatus('Config item `kind` must be set')
```

Comment:
I'm not sure returning status on hooks is rich enough.  For example, what if an `update_status` hook checks relations and finds two are broken - how does that get surfaced?

Reply:
I don't think `update_status` hooks should really be used at all, as they're semi-broken. However if you really wanted to use one like that, you could just do `return self.app, BlockedStatus("There are two things broken: foo, bar")`. The handling over in the operator framework would still take care of updating the status when `update_status` hook ran again, and did ``return self.app, BlockedStatus("There is one thing broken: foo")` instead.

Also, one advantage of this method is if it doesn't work for a particular use case, a charm author can `return None` and continue to manually set `self.model.unit.status`, as they can currently do.